src/serviceContext.js

```diff
@@ -0,0 +1,129 @@
+const crypto = require('crypto');
+const KeyJar = require('../nodeOIDCMsg/src/oicMsg/keystore/KeyJar').KeyJar;
+
+/** 
+ * This class keeps information that a client needs to be able to talk
+ * to a server. Some of this information comes from configuration and some
+ * from dynamic provider info discovery or client registration.
+ * But information is also picked up during the conversation with a server.
+ */
+class ServiceContext {
+  /**
+   * @param {KeyJar} keyjar OIDCMsg KeyJar instance that contains the RP signing and encyrpting keys
+   * @param {Object<string, string>} config Client configuration
+   * @param {Object<string, string>} params Other attributes that might be needed
+   */
+  constructor(keyjar, config, params) {
+    this.keyjar = keyjar || new KeyJar();
+    this.providerInfo = {};
+    this.registrationResponse = {};
+    this.kid = {'sig': {}, 'enc': {}};
+    this.config = config || {};
+    let defaultVal = '';
+
+    if (params) {
+      for (var i = 0; i < Object.keys(params).length; i++){
+        let key = Object.keys(params)[i];
+        let val = params[key];
+        this[key] = val;
+      }
+    }
+
+    this.client_id = this.config['client_id'] || defaultVal;
+    this.issuer = this.config['issuer'] || defaultVal;
+    this.client_secret = this.config['client_secret'] || defaultVal;
+    this.setclient_secret(this.client_secret);    
+    this.base_url = this.config['base_url'] || defaultVal;
+    this.request_dir = this.config['requests_dir'] || defaultVal;
+
+    defaultVal = {}
+    this.allow = this.config['allow'] || defaultVal;
+    this.client_prefs = this.config['client_preferences'] || defaultVal;
+    this.behavior = this.config['behaviour'] || defaultVal;
+    this.provider_info = this.config['provider_info'] || defaultVal;
+
+    try {
```

Comment:
This could be simplified to `this.redirectUris = this.config['redirect_uris'] || [null];`.